pilopt/src/qse_opt.rs

```diff
@@ -76,51 +99,93 @@ fn try_algebraic_to_quadratic_symbolic_expression<T: FieldElement>(
             }
         }
         AlgebraicExpression::UnaryOperation(AlgebraicUnaryOperation { op, expr }) => match op {
-            AlgebraicUnaryOperator::Minus => -try_algebraic_to_quadratic_symbolic_expression(expr)?,
+            AlgebraicUnaryOperator::Minus => -algebraic_to_quadratic_symbolic_expression(expr),
         },
-        AlgebraicExpression::PublicReference(_) | AlgebraicExpression::Challenge(_) => return None,
-    })
+    }
 }
 
 fn quadratic_symbolic_expression_to_algebraic<T: FieldElement>(
-    expr: &QuadraticSymbolicExpression<T, AlgebraicReference>,
+    expr: &QuadraticSymbolicExpression<T, Variable>,
 ) -> AlgebraicExpression<T> {
     let (quadratic, linear, constant) = expr.elements();
     quadratic
         .iter()
         .map(|(l, r)| {
-            quadratic_symbolic_expression_to_algebraic(l)
-                * quadratic_symbolic_expression_to_algebraic(r)
+            let l = quadratic_symbolic_expression_to_algebraic(l);
+            let r = quadratic_symbolic_expression_to_algebraic(r);
+            let (l, l_negated) = extract_negation_if_possible(l);
+            let (r, r_negated) = extract_negation_if_possible(r);
+            match (l_negated, r_negated) {
+                (false, false) => l * r,
+                (false, true) => -(l * r),
+                (true, false) => -(l * r),
+                (true, true) => l * r,
+            }
```

Comment:
Hmm, shouldn't we do simplifications like this in `AlgebraicExpression::mul`?

Reply:
AlgebraicExpression is such a widely used struct that I wouldn't do optimizations there.